python/tvm/topi/unique.py

```diff
@@ -0,0 +1,181 @@
+    output[0] = data[data.shape[0] - 1] + int32(1)
+    return output
+
+
+@hybrid.script
+def _calc_unique_sorted(data, argsorted_indices, inc_scan):
+    unique_elements = output_tensor(data.shape, data.dtype)
+    indices = output_tensor(data.shape, "int32")
+    for i in parallel(data.shape[0]):
+        indices[argsorted_indices[i]] = inc_scan[i]
+        if i == 0 or inc_scan[i] != inc_scan[i - 1]:
+            unique_elements[inc_scan[i]] = data[argsorted_indices[i]]
+    return unique_elements, indices
+
+
+@hybrid.script
+def _calc_unique_sorted_with_counts(data, argsorted_indices, inc_scan):
+    unique_elements = output_tensor(data.shape, data.dtype)
+    indices = output_tensor(data.shape, "int32")
+    counts = output_tensor(data.shape, "int32")
+    for i in parallel(data.shape[0]):
+        counts[i] = int32(0)
+    for i in parallel(data.shape[0]):
+        indices[argsorted_indices[i]] = inc_scan[i]
+        if i == 0 or inc_scan[i] != inc_scan[i - 1]:
+            unique_elements[inc_scan[i]] = data[argsorted_indices[i]]
+    for i in range(data.shape[0]):
+        counts[inc_scan[i]] += int32(1)
+    return unique_elements, indices, counts
+
+
+@hybrid.script
+def _calc_first_occurence(argsorted_indices, inc_scan):
+    first_occurence = output_tensor(argsorted_indices.shape, "int32")
+    for i in parallel(argsorted_indices.shape[0]):
+        first_occurence[i] = argsorted_indices.shape[0]
+    for i in parallel(argsorted_indices.shape[0]):
+        if i == 0 or inc_scan[i] != inc_scan[i - 1]:
+            first_occurence[inc_scan[i]] = argsorted_indices[i]
+    return first_occurence
+
+
+@hybrid.script
+def _calc_unique_unsorted(data, argsorted_indices, inc_scan, index_converter):
+    unique_elements = output_tensor(data.shape, data.dtype)
+    indices = output_tensor(data.shape, "int32")
+    for i in parallel(data.shape[0]):
+        new_unique_idx = index_converter[inc_scan[i]]
+        new_data_idx = argsorted_indices[i]
+        indices[new_data_idx] = new_unique_idx
+        if i == 0 or inc_scan[i] != inc_scan[i - 1]:
+            unique_elements[new_unique_idx] = data[new_data_idx]
+    return unique_elements, indices
+
+
+@hybrid.script
+def _calc_unique_unsorted_with_counts(data, argsorted_indices, inc_scan, index_converter):
+    unique_elements = output_tensor(data.shape, data.dtype)
+    indices = output_tensor(data.shape, "int32")
+    counts = output_tensor(data.shape, "int32")
+    for i in parallel(data.shape[0]):
+        counts[i] = int32(0)
+    for i in parallel(data.shape[0]):
+        new_unique_idx = index_converter[inc_scan[i]]
+        new_data_idx = argsorted_indices[i]
+        indices[new_data_idx] = new_unique_idx
+        if i == 0 or inc_scan[i] != inc_scan[i - 1]:
+            unique_elements[new_unique_idx] = data[new_data_idx]
+    for i in range(data.shape[0]):
+        idx = index_converter[inc_scan[i]]
+        counts[idx] += int32(1)
+    return unique_elements, indices, counts
+
```

Comment:
I'm definitely seeing duplicated code here, but that's one I reason I don't like hybridscript: It doesn't have any abstraction like function

Reply:
I have changed hybrid script to ir builder similar to the GPU implementation.

Reply:
It looks much cleaner now, very nice. Thanks.